### PR TITLE
Add thing API test file

### PR DESCRIPTION
**Pull request by liaojason2softchef**
Fixes #

### CODE REVIEW

File: lambda-assets/things/delete-thing/app.ts [hunk 1]
```diff
@@ -5,17 +5,23 @@ export async function handler(event: { [key: string]: any }) {
   const request = new Request(event);
   const response = new Response();
   try {
+    console.log(request)
     const iotClient = new IoTClient({});
     const thing = await iotClient.send(
       new DeleteThingCommand({
-        thingName: request.parameter('thingName'),
+        thingName: request.input('thingName'),
```
Comment (MinCheTsai) [inline lambda-assets/things/delete-thing/app.ts:12]: from parameter?

clean code, no console.log

File: lambda-assets/things/list-things/app.ts [hunk 2]
```diff
@@ -11,9 +11,7 @@ export async function handler(event: { [key: string]: any }) {
         nextToken: request.get('nextToken', undefined),
       }),
     );
-    return response.json({
-      things,
-    });
+    return response.json(things);
```
Comment (MinCheTsai) [inline lambda-assets/things/list-things/app.ts:14]: request.get('nextToken', undefined) 這裡會出bug?

@Kerwin-Jhong

File: lambda-assets/things/update-thing/app.ts [hunk 3]
```diff
@@ -5,16 +5,29 @@ export async function handler(event: { [key: string]: any }) {
...
+      };
+    });
+    if (validated.error) {
+      return response.error(validated.details, 422);
+    }
     const iotClient = new IoTClient({});
     const thing = await iotClient.send(
       new UpdateThingCommand({
-        thingName: request.parameter('thingName'),
+        thingName: request.input('thingName'),
```
Comment (MinCheTsai) [inline lambda-assets/things/update-thing/app.ts:19]: from parameter

File: test/lambda/thing.test.ts [hunk 4]
```diff
@@ -1,120 +1,224 @@
...
+  mockClient,
+  AwsError,
+} from 'aws-sdk-client-mock';import * as createThing from '../../lambda-assets/things/create-thing/app';
+import * as deleteThing from '../../lambda-assets/things/delete-thing/app';
+import * as getThing from '../../lambda-assets/things/get-thing/app';
+import * as listThings from '../../lambda-assets/things/list-things/app';
+import * as updateThing from '../../lambda-assets/things/update-thing/app';
 
-// const iotClientMock = mockClient(IoTClient);
+//const iotClientMock = mockClient(IoTClient);
```
Comment (MinCheTsai) [inline test/lambda/thing.test.ts:19]: clean code

File: test/lambda/thing.test.ts [hunk 5]
```diff
@@ -1,120 +1,224 @@
...
-// import * as createThing from '../../lambda-assets/things/create-thing/app';
-// import * as deleteThing from '../../lambda-assets/things/delete-thing/app';
-// import * as updateThing from '../../lambda-assets/things/update-thing/app';
-// import * as listThings from '../../lambda-assets/things/list-things/app';
-// import * as getThing from '../../lambda-assets/things/get-thing/app';
+import {
+  IoTClient,
+  CreateThingCommand,
+  UpdateThingCommand,
+  //DescribeThingCommand,
```
Comment (MinCheTsai) [inline test/lambda/thing.test.ts:5]: No mock?

File: test/lambda/thing.test.ts [hunk 6]
```diff
@@ -1,120 +1,224 @@
...
-  // const body = JSON.parse(response.body);
-  // expect(response.statusCode).toEqual(404);
+test('Update thing with invalid thingName expect failure', async () => {
+  const iotClientMock = mockClient(IoTClient);
+  iotClientMock.on(UpdateThingCommand, {
+    thingName: '(((ﾟДﾟ;)))',
+  }).rejects(expected.invalidThingError);
+  const response = await updateThing.handler({
+    body: {
+      thingName: '(((ﾟДﾟ;)))',
```
Comment (MinCheTsai) [inline test/lambda/thing.test.ts:189]: 統一thingName到expected.thing.thingName

File: lambda-assets/things/list-things/app.ts [hunk 7]
```diff
@@ -1,18 +1,19 @@
...
-import { Request, Response } from '../../utils';
+import { Request, Response } from '@softchef/lambda-events';
 
 export async function handler(event: { [key: string]: any }) {
   const request = new Request(event);
   const response = new Response();
   try {
     const iotClient = new IoTClient({});
-    const things = await iotClient.send(
+    const {things: things, nextToken: nextToken} = await iotClient.send(
```
Comment (MinCheTsai) [inline lambda-assets/things/list-things/app.ts:9]: ??

lint?

File: lambda-assets/things/update-thing/app.ts [hunk 8]
```diff
@@ -1,20 +1,40 @@
...
       new UpdateThingCommand({
         thingName: request.parameter('thingName'),
+        thingTypeName: request.input('thingTypeName'),
+        attributePayload: request.input('attributePayload'),
+        expectedVersion: request.input('expectedVersion'),
+        removeThingType: request.input('removeThingType'),
       }),
     );
     return response.json({
+      updated: true,
```
Comment (MinCheTsai) [inline lambda-assets/things/update-thing/app.ts:30]: 無需回傳thing

File: lambda-assets/things/update-thing/app.ts [hunk 9]
```diff
@@ -1,20 +1,40 @@
...
+      };
+    });
+    if (validated.error) {
+      return response.error(validated.details, 422);
+    }
     const iotClient = new IoTClient({});
     const thing = await iotClient.send(
       new UpdateThingCommand({
         thingName: request.parameter('thingName'),
+        thingTypeName: request.input('thingTypeName'),
```
Comment (MinCheTsai) [inline lambda-assets/things/update-thing/app.ts:23]: @Kerwin-Jhong 請協助確認 這樣API能用？

File: test/lambda/thing.test.ts (hunk 5 above)
Comment (liaojason2softchef) [reply test/lambda/thing.test.ts]: DescribeThingCommand 下面有宣告過了

File: lambda-assets/things/update-thing/app.ts (hunk 9 above)
Comment (liaojason2softchef) [reply lambda-assets/things/update-thing/app.ts]: 確認可

File: lambda-assets/things/update-thing/app.ts [hunk 10]
```diff
@@ -1,20 +1,43 @@
...
+        removeThingType: joi.boolean().allow(null),
+      };
+    });
+    if (validated.error) {
+      return response.error(validated.details, 422);
+    }
     const iotClient = new IoTClient({});
-    const thing = await iotClient.send(
+    let parameters: { [key: string]: any } = {};
+    const attributes = ['thingTypeName', 'attributePayload','expectedVersion','removeThingType']
```
Comment (MinCheTsai) [inline lambda-assets/things/update-thing/app.ts:21]: lint

File: lambda-assets/things/get-thing/app.ts [hunk 11]
```diff
@@ -1,5 +1,5 @@
 import { IoTClient, DescribeThingCommand } from '@aws-sdk/client-iot';
-import { Request, Response } from '../../utils';
+import { Request, Response } from '@softchef/lambda-events';
```
Comment (MinCheTsai) [inline lambda-assets/things/get-thing/app.ts:2]: if 404?

File: test/lambda/thing.test.ts [hunk 12]
```diff
@@ -1,120 +1,248 @@
...
+      thingTypeName: expected.updateThing.thingTypeName,
+      attributePayload: expected.updateThing.attributePayload,
+      expectedVersion: expected.updateThing.expectedVersion,
+      removeThingType: expected.updateThing.removeThingType,
+    },
+  });
+  const body = JSON.parse(response.body);
+  expect(response.statusCode).toEqual(200);
+  expect(body.updated).toEqual(true);
+  iotClientMock.restore;
```
Comment (MinCheTsai) [inline test/lambda/thing.test.ts:186]: ?

File: test/lambda/thing.test.ts [hunk 13]
```diff
@@ -1,120 +1,248 @@
...
+  });
+  const response = await getThing.handler({
+    pathParameters: {
+      thingName: expected.thing.thingName,
+    },
+  });
+  const body = JSON.parse(response.body);
+  expect(body.thing).toEqual(expected.thing);
+  expect(response.statusCode).toEqual(200);
+  iotClientMock.restore();
```
Comment (MinCheTsai) [inline test/lambda/thing.test.ts:125]: Test: "Get thing with invalid thingName expect failure"

**Review by MinCheTsai — APPROVED** [review]

**Review by Kerwin-Jhong — APPROVED** [review]

**Pull request merged**